consensus/parlia/parlia.go

```diff
@@ -1229,6 +1663,79 @@ func (p *Parlia) applyTransaction(
 	return nil
 }
 
+func (p *Parlia) SetVotePool(votePool consensus.VotePool) {
+	p.votePool = votePool
+}
+
+func (p *Parlia) GetHighestJustifiedHeader(chain consensus.ChainHeaderReader, header *types.Header) *types.Header {
+	if chain == nil || header == nil {
+		return nil
+	}
+
+	// There won't any attestation in first two blocks, return root.
+	if header.Number.Uint64() < 2 {
+		return chain.GetHeaderByNumber(0)
+	}
+
+	parent := chain.GetHeaderByHash(header.ParentHash)
+	if parent == nil {
+		return nil
+	}
+	snap, err := p.snapshot(chain, parent.Number.Uint64(), parent.Hash(), nil)
+	if err != nil {
+		log.Error("Unexpected error when getting snapshot",
+			"error", err, "blockNumber", parent.Number.Uint64(), "blockHash", parent.Hash())
+		return nil
+	}
+
+	// If there is no vote justified block, then return root.
+	if snap.Attestation == nil {
+		return chain.GetHeaderByNumber(0)
+	}
+	// Return naturally justified block.
+	if snap.Number-snap.Attestation.TargetNumber > naturallyJustifiedDist {
+		return FindAncientHeader(parent, naturallyJustifiedDist, chain, nil)
+	}
+	//Return latest vote justified block.
+	return chain.GetHeaderByHash(snap.Attestation.TargetHash)
+}
+
+func (p *Parlia) GetHighestFinalizedNumber(chain consensus.ChainHeaderReader, header *types.Header) uint64 {
+	if !chain.Config().IsLynn(header.Number) {
+		return 0
+	}
+	if chain == nil || header == nil || header.Number.Uint64() < 2 {
+		return 0
+	}
+
+	parent := chain.GetHeaderByHash(header.ParentHash)
+	if parent == nil {
+		return 0
+	}
+
+	snap, err := p.snapshot(chain, parent.Number.Uint64(), parent.Hash(), nil)
+	if err != nil {
+		log.Error("Unexpected error when getting snapshot",
+			"error", err, "blockNumber", parent.Number.Uint64(), "blockHash", parent.Hash())
+		return 0
+	}
+
+	for snap.Attestation != nil {
```

Comment:
Did not get the idea here. why it's not `return snap.Attestation.TargetNumber`?

Reply:
Because the source block is the highest justified block before the target block, it has been justified, once we get the target block's attestation, the target block has been justified. When a block's attestation's targetNumber == sourceNumber + 1, that means there are two consecutive justified blocks, and the previous justified block is finalized according to the finalize rule.